2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs

```diff
@@ -0,0 +1,114 @@
+    }
+
+    public override void UpdateState(RockBossStateManager RockBoss)
+    {
+        if(attackTime <= 0)
+        {
+            animator.SetBool("Beaming", false);
+            animator.SetBool("Shooting", false);
+            animator.SetBool("Charging", false);
+            animator.SetBool("Spinning", false);
+            animator.SetBool("Slamming", false);
+            RockBoss.SwitchState(RockBoss.IdleState);
+        }
+
+        attackTime -= Time.deltaTime;
+    }
+
+    public override void OnCollisionEnter2D(RockBossStateManager RockBoss, Collision2D other)
+    {
+        
+    }
+    
+    public override void OnTriggerStay2D(RockBossStateManager RockBoss, Collider2D other) 
+    {
+
+    }
+
+    //done in animation events
+    public override void EventTrigger(RockBossStateManager RockBoss)
+    {
+        Vector2 knockbackDirection = (Vector2)(RockBoss.transform.position - RockBoss.attackPointX.position).normalized;
+        LayerMask mask = LayerMask.GetMask("Player");
+        Collider2D[] colliders = Physics2D.OverlapCapsuleAll(RockBoss.attackPointX.position, new Vector2(RockBoss.attackRange * 2, RockBoss.attackHeight), CapsuleDirection2D.Vertical, mask);
+
+        foreach (Collider2D collider in colliders)
+        {
+            if (collider.CompareTag("Player"))
+            {
+                PlayerController playerScript = collider.GetComponent<PlayerController>();
+                if (returnAttackType == 0) //beam attack
+                {
+                    playerScript.dealDamage(15);
+                    collider.GetComponent<Rigidbody2D>().AddForce(-knockbackDirection * 25, ForceMode2D.Impulse);
+                    collider.GetComponent<Animator>().SetTrigger("Hit");
+                }
+                else if (returnAttackType == 1) //shooting attack
+                {
+                    playerScript.dealDamage(12);
+                    collider.GetComponent<Rigidbody2D>().AddForce(-knockbackDirection * 20, ForceMode2D.Impulse);
+                    collider.GetComponent<Animator>().SetTrigger("Hit");
+                }
+                else if (returnAttackType == 2) //charging
+                {
+                    playerScript.dealDamage(10);
+                    collider.GetComponent<Rigidbody2D>().AddForce(-knockbackDirection * 40, ForceMode2D.Impulse);
+                    collider.GetComponent<Animator>().SetTrigger("Hit");
+                }
+                else if (returnAttackType == 3) //slamming
+                {
+                    playerScript.dealDamage(10);
+                    collider.GetComponent<Rigidbody2D>().AddForce(-knockbackDirection * 5, ForceMode2D.Impulse);
+                    collider.GetComponent<Animator>().SetTrigger("Hit");
+                    playerScript.ReduceMoveSpeed(35, 1.5f);
+                }
+                
+            }
+        }
+        
+    }
+    public override void TakeDamage(RockBossStateManager RockBoss)
```

Comment:
A big problem here, and this goes for all the states, is that I believe bosses should not have hitstun. They could just be permanently hit over and over again and just have their attack get cancelled. I really do thing its best to just not have any hitstun. Simply moving over to the Hit State would interrupt his attack, so just dont have him go over to the hitstate, and instead when calling TakeDamage, just check if the boss' HP < 0.